Documentation updates

kamillf opened the pull request:
Here's an initial draft of the modified readme for Simpler (minus anything test-case related, which includes Name and Stats properties). Before I go further, please let me know what you think.

kamillf committed:
Initial commit

Checking on graphics

kamillf committed:
Add Task Info

kamillf committed:
Test Linkages

kamillf committed:
Try links again

kamillf committed:
Fixing Links

kamillf committed:
All but test cases

gregoryjscott commented on README.md:
```diff
@@ -1,259 +1,237 @@
-        Console.WriteLine(String.Format("{0} finished.", task.Name));
-    }
+1. [Create a Task class] (#creating_tasks). 
+2. [Instantiate the Task] (#instantiating_tasks) using the Task.New() method. 
```
Needs a 3. Execute the task.

And you might need to pass some input, and then do something with the output, but that probably doesn't need to be mentioned here.

gregoryjscott commented on README.md:
```diff
@@ -1,259 +1,237 @@
-            Ask.Execute();
-        }
+>Example: A Task that parses an XML file containing projects might be called *ParseProjectsXML*. But it shouldn’t be called *XMLProjectParser* or *XMLProject*.
 
```
I like this example. 

I'll point out one nit-picky thing about the naming - typically acronyms like XML are not all caps in class names. The preferred name would be `ParseProjectsXml`. Here's some examples from the .NET framework's `System.Xml` namespace http://msdn.microsoft.com/en-us/library/system.xml.aspx.

gregoryjscott commented on README.md:
```diff
@@ -1,259 +1,237 @@
+>- [`Name`] (#name) property
+- [`Stats`] (#stats) property
+- Static `Task.New<TTask>()` method, which is a factory method for [instantiating tasks] (#instantiating_tasks) 
+
```
Since `Task.New<>` is a static method it isn't correct to say the the other 3 classes inherit this method. It's only available from the Task class itself.

FYI, I've considered moving the New method to a new class called Factory. The code would read like this.

``` c#
var myTask = Factory.New<MyTask>();
```

That might help clear up the purpose of the Task class itself.

gregoryjscott commented on README.md:
```diff
@@ -1,259 +1,237 @@
-        }
+###<a name="instantiating_tasks"></a>Instantiating Tasks
+
+When you have [created at least one Task](#creating_tasks), you can instantiate a Task using the `Task.New()` method.
```
"... you can instantiate _the_ Task using ..."

To my ears, _the_ makes it clearer that you can instantiate _the_ Task you created, not just _a_ Task.

gregoryjscott commented on README.md:
```diff
@@ -1,259 +1,237 @@
-Exactly. I seriously hope to create some proper documentation at some point, but the coding is so much more fun.
 
-###Acknowledgments
+##Acknowledgments
```
Initially I was adding anyone who was using Simpler and giving me feedback to this acknowledgments list. However, I haven't been good about updating it lately. Not sure what to do about this section.

gregoryjscott commented:
By all means, keep going!

To show the support for testing, we could show how a Task class and a Task test class would be developed step by step using a [TDD](http://en.wikipedia.org/wiki/Test-driven_development) workflow. I enjoy the TDD workflow and Simpler makes it pretty easy.

kamillf committed:
* Addressed most of Greg’s comments
* Added in testing information, including TDD workflow example
* Fixed a couple examples (hopefully they’re legit now)
* Added section related to EventsAttribute
[ ] Still need to finish off Stats and Name Properties, but that’s a
thing for another day….

kamillf committed:
Intermediate check-in

* Added an edited contributing document
* Changed the Simpler diagram to reflect the ability to organize the
Tasks and Models
* Started updating the examples in the ReadMe to all be about wildlife
sightings (a few are still left)
* Removed acknowledgements so Greg can latter add a Hall of Fame instead
* Added content for the Name and Stats properties
* Did some re-organizing

kamillf commented:
On Contributing, should CI should be referenced in the section about how to make a contribution?

kamillf committed:
Final for today

* A few more organizational changes
* Updated EventsAttribute example to use a wildlife sample

kamillf commented:
The examples in the Testing section haven't been updated to match the others. But I don't have the brain power right now to do more. If you can come up with examples to use, that would be wonderful.

kamillf commented on README.md:
```diff
@@ -1,275 +1,511 @@
 
         // Assert
-        Assert.That(task.Out.Stuff.Length, Is.EqualTo(9));
+        Assert.That(task.Stats.ExecuteDurations, Is.LessThan(3));
```
What are the units on ExecuteDurations? 3 may or may not be realistic....

gregoryjscott replied:
Don't bother changing this, but FYI...

ExecuteDurations is a list of `TimeSpan` (.NET Framework) classes that have methods for getting time in various units. If you're expecting the task to be executed multiple times then your test probably should have something like this instead.

``` c#
Assert.That(task.Stats.ExecuteDurations
    .Average(ed => ed.TotalSeconds), Is.LessThan(3));
```

> This was not compiled.

gregoryjscott commented:
No problem @kamillf, I'll work on the examples. I have an idea for a subject - we'll see if it works.

gregoryjscott commented:
I'm going to merge this and then open a new PR that replaces the existing examples.
